tests/serverpod_test_server/test_integration/test_tools/reset_test_session_test.dart

```diff
@@ -0,0 +1,100 @@
+import 'package:serverpod/serverpod.dart';
+import 'package:test/test.dart';
+
+import 'serverpod_test_tools.dart';
+
+void main() {
+  group('Given resetTestSessions set to afterEach', () {
+    withServerpod(
+      (endpoints, session) {
+        group('when copying the session in setUpAll', () {
+          late TestSession newSession;
+          setUpAll(() async {
+            newSession = await session.copyWith(
+                getAuthenticationInfo: () => AuthenticationInfo(123, {}));
+          });
+
+          test('then the first test has the modified newSession', () async {
+            expect(newSession.authenticationInfo!.userId, 123);
+          });
+
+          test('then the second test has the reset newSession', () async {
+            expect(newSession.authenticationInfo, isNull);
+          });
+        });
+
+        group('when copying the session in setUp', () {
+          late TestSession newSession;
+          setUp(() async {
+            newSession = await session.copyWith(
+                getAuthenticationInfo: () => AuthenticationInfo(123, {}));
+          });
+
+          test('then the first test has the modified newSession', () async {
+            expect(newSession.authenticationInfo!.userId, 123);
+          });
+
+          test('then the second test has the modified newSession', () async {
+            expect(newSession.authenticationInfo!.userId, 123);
+          });
+        });
+
+        group('when copying the session in the first test', () {
+          late TestSession newSession;
+
+          test('then the first test has the modified newSession', () async {
+            newSession = await session.copyWith(
+                getAuthenticationInfo: () => AuthenticationInfo(123, {}));
+
+            expect(newSession.authenticationInfo!.userId, 123);
+          });
+
+          test('then the second test has the reset newSession', () async {
+            expect(newSession.authenticationInfo, isNull);
+          });
+        });
+      },
+      resetTestSessions: ResetTestSessions.afterEach,
```

Comment:
I think it would be really valuable to have these at the top of the test instead of the bottom. Reading test cases is naturally done top-down and having these semantics defined at the end seems to risk a gotcha everytime they're not the default.. 

In order to get around the problem with default params and named params having to be last, a couple of approaches could be considered:

- Leaving `withServerpod()` as-is, but adding wrapper functions with clear names that apply these settings, such as `withServerpodResetEach()`. The drawback is somewhat weird and long names and if there are many common permutations this gets unwieldy.

- Making the main argument function (endpoints, session){} a named parameter with a simple name like func. Not much more to write but clear and flexible:
```
    withServerpod(
      resetTestSessions: ResetTestSessions.afterEach,
      runMode: ServerpodRunMode.production,
      func: (endpoints, session) {
        ...
      },
    )
```

Reply:
This is great feedback and those are valid alternatives. The two main arguments for the current setup are:
- I think the 95% percentile use case will have _no_ options; so "no options" is the case that should be optimized for and should not contain unnecessary "boiler plate".
- Dart `test` and `group` have their options last just like this, and keeping the same structure will provide familiarity for the user and hopefully be more intuitive.
- (We're specifying `runMode` in each test, but this should actually be extremely rare in the wild, since going forward there shouldn't be many use cases for not using `test` mode. Our `serverpod_test_server.dart` was currently using `prod` so I opted here to not use `test` in some tests and `prod` in all current tests).

Would be great to get input from @Isakdl and @vlidholt as well!

Reply:
I agree that you get a better overview with the option in the top, but as you point out it is a good idea to mirror the structure that already exists in the dart test tooling. I'm in favor of keeping it as is for this reason only.